content_handler/runtime_holder.cc

```diff
@@ -103,6 +111,57 @@ void RuntimeHolder::Init(
   outgoing_services_ = std::move(outgoing_services);
 
   InitRootBundle(std::move(bundle));
+
+  const uint8_t* vm_snapshot_data;
+  const uint8_t* vm_snapshot_instr;
+  const uint8_t* default_isolate_snapshot_data;
+  const uint8_t* default_isolate_snapshot_instr;
+  if (!Dart_IsPrecompiledRuntime()) {
+    vm_snapshot_data = ::kDartVmSnapshotData;
+    vm_snapshot_instr = ::kDartVmSnapshotInstructions;
+    default_isolate_snapshot_data = ::kDartIsolateCoreSnapshotData;
+    default_isolate_snapshot_instr = ::kDartIsolateCoreSnapshotInstructions;
+  } else {
+    std::vector<uint8_t> dylib_blob;
+    if (!asset_store_->GetAsBuffer(kDylibKey, &dylib_blob)) {
+      FTL_LOG(ERROR) << "Failed to extract app dylib";
+      return;
+    }
+
+    mx::vmo dylib_vmo;
+    if (!mtl::VmoFromVector(dylib_blob, &dylib_vmo)) {
+      FTL_LOG(ERROR) << "Failed to load app dylib";
+      return;
+    }
+
+    dlerror();
+    dylib_handle_ = dlopen_vmo(dylib_vmo.get(), RTLD_LAZY);
+    if (dylib_handle_ == nullptr) {
+      FTL_LOG(ERROR) << "dlopen failed: " << dlerror();
+      return;
+    }
+    vm_snapshot_data = reinterpret_cast<const uint8_t*>(
+        dlsym(dylib_handle_, "_kDartVmSnapshotData"));
+    vm_snapshot_instr = reinterpret_cast<const uint8_t*>(
+        dlsym(dylib_handle_, "_kDartVmSnapshotInstructions"));
+    default_isolate_snapshot_data = reinterpret_cast<const uint8_t*>(
+        dlsym(dylib_handle_, "_kDartIsolateSnapshotData"));
+    default_isolate_snapshot_instr = reinterpret_cast<const uint8_t*>(
+        dlsym(dylib_handle_, "_kDartIsolateSnapshotInstructions"));
+  }
+
+  // TODO(rmacnak): We should generate the AOT vm snapshot separately from
+  // each app so we can initialize before receiving the first app bundle.
+  static std::atomic_bool first_app(true);
```

Comment:
Even if there were many threads, wouldn't you need to do a compare and exchange?